rama-http/src/headers/x_robots_tag/robots_tag.rs

```diff
@@ -0,0 +1,102 @@
+use crate::headers::util::value_string::HeaderValueString;
+use crate::headers::x_robots_tag::custom_rule::CustomRule;
+use crate::headers::x_robots_tag::max_image_preview_setting::MaxImagePreviewSetting;
+use crate::headers::x_robots_tag::robots_tag_builder::RobotsTagBuilder;
+
+macro_rules! getter_setter {
+    ($field:ident, $type:ty) => {
+        paste::paste! {
+            pub fn [<$field>](&self) -> $type {
+                self.[<$field>]
+            }
+
+            pub fn [<set_ $field>](&mut self, [<$field>]: $type) -> &mut Self {
+                self.[<$field>] = [<$field>];
+                self
+            }
+        }
+    };
+
+    ($field:ident, $type:ty, optional) => {
+        paste::paste! {
+            pub fn [<$field>](&self) -> Option<&$type> {
+                self.[<$field>].as_ref()
+            }
+
+            pub fn [<set_ $field>](&mut self, [<$field>]: $type) -> &mut Self {
+                self.[<$field>] = Some([<$field>]);
+                self
+            }
+        }
+    };
+
+    ($field:ident, $type:ty, vec) => {
+        paste::paste! {
+            pub fn [<$field>](&self) -> &Vec<$type> {
+                &self.[<$field>]
+            }
+
+            pub fn [<add_ $field>](&mut self, [<$field>]: $type) -> &mut Self {
+                self.[<$field>].push([<$field>]);
+                self
+            }
+        }
+    };
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Default)]
+pub struct RobotsTag {
+    bot_name: Option<HeaderValueString>,
+    all: bool,
+    no_index: bool,
+    no_follow: bool,
+    none: bool,
+    no_snippet: bool,
+    index_if_embedded: bool,
+    max_snippet: u32,
+    max_image_preview: Option<MaxImagePreviewSetting>,
+    max_video_preview: Option<u32>,
+    no_translate: bool,
+    no_image_index: bool,
+    unavailable_after: Option<chrono::DateTime<chrono::Utc>>, // "A date must be specified in a format such as RFC 822, RFC 850, or ISO 8601."
+    // custom rules
+    no_ai: bool,
+    no_image_ai: bool,
+    spc: bool,
+    custom_rules: Vec<CustomRule>,
+}
+
+impl RobotsTag {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn with_bot_name(bot_name: Option<HeaderValueString>) -> Self {
+        Self {
+            bot_name,
+            ..Default::default()
+        }
+    }
+
+    pub fn builder() -> RobotsTagBuilder {
+        RobotsTagBuilder::new()
+    }
+
+    getter_setter!(bot_name, HeaderValueString, optional);
```

Comment:
Why are we having a builder as well as a setter in the struct? At that point what additional guarantees or reasons do we have to even have the builder?